### PR TITLE
Fix 460: When sample animation mode is used, bake all bones

### PR DESCRIPTION
**Pull request by julienduroure**
When sample animation mode is used, all bones are baked, not only bones with keyframes.
This will allow to export complex rigs, where deformation bones are not animated. This is how fbx export works, for example.

Tested on some various rigs, including a bouncing ball,and a simple character with IK leg / foot roll.
[460.zip](https://github.com/KhronosGroup/glTF-Blender-IO/files/3194998/460.zip)

### CODE REVIEW

**Comment by julienduroure** [comment]
Seems ready to be tested @emackey @donmccurdy

**Comment by emackey** [comment]
Unfortunately I'm not the right person to be reviewing skinning/bone/rigging code.  @donmccurdy or @pjoe or @Cyp, would you guys have any time to look at this?  Feedback from anyone else is welcome too of course.

File: addons/io_scene_gltf2/blender/exp/gltf2_blender_gather_animation_channels.py [hunk 1]
```diff
@@ -35,8 +35,21 @@ def gather_animation_channels(blender_action: bpy.types.Action,
...
+        for chans in groups:
+            ranges = [channel.range() for channel in chans]
+            if bake_range_start is None:
+                bake_range_start = min([channel.range()[0] for channel in chans])
+            else:
+                bake_range_start = min(bake_range_start, min([channel.range()[0] for channel in chans]))
+            if bake_range_end is None:
+                bake_range_end = max([channel.range()[1] for channel in chans])
+            else:
+                bake_range_rend = max(bake_range_end, max([channel.range()[1] for channel in chans]))
```
Comment (Cyp) [inline addons/io_scene_gltf2/blender/exp/gltf2_blender_gather_animation_channels.py]: `bake_range_rend` typo?

**Comment by donmccurdy** [comment]
I'll review it this week.

(and thank you @julienduroure for the PR and @Cyp for the review, much appreciated!)

File: addons/io_scene_gltf2/blender/exp/gltf2_blender_gather_animation_channels.py (hunk 1 above)
Comment (julienduroure) [reply addons/io_scene_gltf2/blender/exp/gltf2_blender_gather_animation_channels.py]: Yes, typo. Fixed now

**Comment by pjoe** [comment]
From a quick read it looks pretty good to me :) I will try to do some tests with different files.

Btw. interesting with those python type annotations, though I do feel a bit weird that there is no actual type check - guess I have been doing to much TypeScript lately 😛

File: addons/io_scene_gltf2/blender/exp/gltf2_blender_gather_animation_channel_target.py [hunk 2]
```diff
@@ -24,37 +24,47 @@
 @cached
 def gather_animation_channel_target(channels: typing.Tuple[bpy.types.FCurve],
                                     blender_object: bpy.types.Object,
+                                    bake_bone: typing.Union[str, None],
+                                    bake_channel: typing.Union[str, None],
                                     export_settings
                                     ) -> gltf2_io.AnimationChannelTarget:
```
Comment (donmccurdy) [inline addons/io_scene_gltf2/blender/exp/gltf2_blender_gather_animation_channel_target.py]: Hm, it seems worrying to see a `@cached` function that isn't taking any parameters related to the current NLA track or Action name. This seems like a probable cause of https://github.com/KhronosGroup/glTF-Blender-IO/issues/451 and https://github.com/KhronosGroup/glTF-Blender-IO/issues/461?

**Comment by donmccurdy** [comment]
@julienduroure what export settings should I be using to check the IK leg example? With *Always Sample, or Always Sample + Bake Skinning Constraints, I'm not getting a correct output (although it looks closer than before.

**Comment by donmccurdy** [comment]
Also – None of my skinning examples export correctly on the master branch. I'm happy to do reviews, but I'm also OK if we want to merge PRs more quickly, given that the master branch skinning export is already quite broken as far as I can tell. If there are any skinning examples that do work now, let's upload them somewhere – otherwise my testing things is not very effective. 😔

File: addons/io_scene_gltf2/blender/exp/gltf2_blender_gather_animation_channel_target.py (hunk 2 above)
Comment (julienduroure) [reply addons/io_scene_gltf2/blender/exp/gltf2_blender_gather_animation_channel_target.py]: Yes, you're probably right.
#461 will be my next work, after #426 and #460 will be merged

**Comment by julienduroure** [comment]
@donmccurdy Currently using only *Always sample* because *bake sinning* does nothing in code for now (see #485 for the discussion about that)

You have here 2 examples
[460.zip](https://github.com/KhronosGroup/glTF-Blender-IO/files/3194998/460.zip)

And here is a really complex rig (generate with rigify). Re-import in blender is correct (some glitches maybe linked to only 4 skin weights export ?), but is KO when viewing it in https://gltf-viewer.donmccurdy.com/ :( 
[rigify.zip](https://github.com/KhronosGroup/glTF-Blender-IO/files/3215136/rigify.zip)

**Comment by julienduroure** [comment]
@donmccurdy :

In https://gltf.insimo.com/, animation is better, even if there is still a global orientation issue
And everything seems to be OK in https://sandbox.babylonjs.com/

**Comment by donmccurdy** [comment]
Ok thanks! I've filed https://github.com/mrdoob/three.js/issues/16562 about the animation issue in three.js, we can look into that. We ready to merge this?

**Comment by julienduroure** [comment]
@donmccurdy Before merging, I want to be sure that this behaviour is linked to the right user option in UI. 
My comments and solution proposal are in #485

**Comment by julienduroure** [comment]
Thanks for your feedback. I will manage merging myself later today

**Pull request merged**